src/types/mod.rs

```diff
@@ -320,17 +308,69 @@ impl Bound {
     }
 }
 
+impl fmt::Debug for Bound {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let arc = Arc::new(self.shallow_clone());
+        for data in arc.verbose_pre_order_iter::<NoSharing>() {
+            match (&*data.node, data.n_children_yielded) {
+                (Bound::Free(ref s), _) => f.write_str(s)?,
+                (Bound::Complete(ref comp), _) => fmt::Debug::fmt(comp, f)?,
+                (Bound::Sum(..), 0) | (Bound::Product(..), 0) => f.write_str("(")?,
+                (Bound::Sum(..), 2) | (Bound::Product(..), 2) => f.write_str(")")?,
+                (Bound::Sum(..), _) => f.write_str(" + ")?,
+                (Bound::Product(..), _) => f.write_str(" × ")?,
+            }
+        }
+        Ok(())
+    }
+}
+
 impl fmt::Display for Bound {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match *self {
-            Bound::Free(ref name) => f.write_str(name),
-            Bound::Complete(ref data) => fmt::Display::fmt(data, f),
-            Bound::Sum(ref a, ref b) => write!(f, "({} + {})", a, b),
-            Bound::Product(ref a, ref b) => write!(f, "({} × {})", a, b),
+        let arc = Arc::new(self.shallow_clone());
+        for data in arc.verbose_pre_order_iter::<NoSharing>() {
+            match (&*data.node, data.n_children_yielded) {
+                (Bound::Free(ref s), _) => f.write_str(s)?,
+                (Bound::Complete(ref comp), _) => fmt::Display::fmt(comp, f)?,
+                (Bound::Sum(..), 0) | (Bound::Product(..), 0) => f.write_str("(")?,
+                (Bound::Sum(..), 2) | (Bound::Product(..), 2) => f.write_str(")")?,
+                (Bound::Sum(..), _) => f.write_str(" + ")?,
+                (Bound::Product(..), _) => f.write_str(" × ")?,
+            }
         }
+        Ok(())
     }
 }
 
+impl DagLike for Arc<Bound> {
+    type Node = Bound;
+    fn data(&self) -> &Bound {
+        self
+    }
+
+    fn as_dag_node(&self) -> Dag<Self> {
+        match **self {
+            Bound::Free(..) | Bound::Complete(..) => Dag::Nullary,
+            Bound::Sum(ref ty1, ref ty2) | Bound::Product(ref ty1, ref ty2) => {
+                Dag::Binary(ty1.bound.root().get(), ty2.bound.root().get())
+            }
+        }
+    }
+}
+
+/// Source or target type of a Simplicity expression
+///
+/// This is just a mutex holding the actual data, which defined by the
+/// `TypeInner` type.
```

Comment:
I don't know what commit you're referring to (this code is not in 148dc06) or what code you're referring to, because Github is only showing two lines of context.

Reply:
I cannot find the text "This is just" in 18035d0 either.

In any case, I'll add a commit which updates the doccomment on `Type`, which I infer is what you're referring to.